worker/src/vectorize.rs

```diff
@@ -0,0 +1,319 @@
+    Indexed,
+    /// No indexed metadata will be returned.
+    #[default]
+    None,
+}
+
+#[derive(Debug, Serialize, Hash, PartialEq, Eq)]
+/// Comparison logic/operation to use for metadata filtering.
+///
+/// This list is expected to grow as support for more operations are released.
+pub enum VectorizeVectorMetadataFilterOp {
+    #[serde(rename = "$eq")]
+    Eq,
+    #[serde(rename = "$ne")]
+    Neq,
+}
+
+/// Filter criteria for vector metadata used to limit the retrieved query result set.
+type VectorizeVectorMetadataFilter =
+    HashMap<String, HashMap<VectorizeVectorMetadataFilterOp, serde_json::Value>>;
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct VectorizeQueryOptions {
+    // Default 5, max 100
+    top_k: u8,
+    /// Return vectors from the specified namespace. Default `none`.
+    namespace: Option<String>,
+    /// Return vector values. Default `false`.
+    return_values: bool,
+    /// Return vector metadata. Default `None`.
+    return_metadata: VectorizeMetadataRetrievalLevel,
+    /// Default `none`.
+    filter: Option<VectorizeVectorMetadataFilter>,
+}
+
+impl VectorizeQueryOptions {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn with_top_k(mut self, top_k: u8) -> Self {
+        self.top_k = top_k;
+        self
+    }
+
+    pub fn with_namespace(mut self, namespace: &str) -> Self {
+        self.namespace = Some(namespace.to_owned());
+        self
+    }
+
+    pub fn with_return_values(mut self, return_values: bool) -> Self {
+        self.return_values = return_values;
+        self
+    }
+
+    pub fn with_return_metadata(
+        mut self,
+        return_metadata: VectorizeMetadataRetrievalLevel,
+    ) -> Self {
+        self.return_metadata = return_metadata;
+        self
+    }
+
+    pub fn with_filter_entry<T: Serialize>(
+        mut self,
+        key: &str,
+        op: VectorizeVectorMetadataFilterOp,
+        value: T,
+    ) -> Result<Self> {
+        let mut filter = self.filter.unwrap_or_default();
+        let inner = filter.entry(key.to_owned()).or_default();
+        inner.insert(op, serde_json::to_value(value)?);
+        self.filter = Some(filter);
+        Ok(self)
+    }
+}
+
+impl Default for VectorizeQueryOptions {
+    fn default() -> Self {
+        Self {
+            top_k: 5,
+            namespace: None,
+            return_values: false,
+            return_metadata: VectorizeMetadataRetrievalLevel::None,
+            filter: None,
+        }
+    }
+}
+
+#[derive(Debug, Deserialize)]
+/// Represents a single vector value set along with its associated metadata.
+pub struct VectorizeVectorResult {
+    /// The ID for the vector. This can be user-defined, and must be unique. It should uniquely identify the object, and is best set based on the ID of what the vector represents.
+    pub id: String,
+    /// The vector values.
+    pub values: Option<Vec<f32>>,
```

Comment:
The TS type for values is `VectorFloatArray | number[]`, when would this value ever be `None`?

Reply:
The `query` operation returns a vector of `VectorizeMatchVector` which is composed of this `VectorizeVectorResult`. In the query operation, `values` may or may not be present and that's why this field is optional. I believe this is the right behavior.

In workerd types, we use a Partial to implement this logic: https://workers-types.pages.dev/#VectorizeMatch.